core/src/cluster_slot_state_verifier.rs

```diff
@@ -345,92 +348,120 @@ pub enum ResultingStateChange {
-        state_changes.push(ResultingStateChange::RepairDuplicateConfirmedVersion(
-            duplicate_confirmed_hash,
-        ));
-    } else {
-        // If the versions match, then add the slot to the candidate
-        // set to account for the case where it was removed earlier
-        // by the `on_duplicate_slot()` handler
-        state_changes.push(ResultingStateChange::DuplicateConfirmedSlotMatchesCluster(
-            bank_frozen_hash,
-        ));
     }
 }
 
-fn check_epoch_slots_hash_against_frozen_hash(
+/// Checks the epoch slots hash we observed against our local bank status
+/// Note epoch slots here does not refer to gossip but responses from ancestor_hashes_service
+///
+/// * If `epoch_slots_frozen_hash` matches our local frozen hash do nothing
+/// * If `slot` has not yet been replayed and is not pruned do nothing
+///
+/// Dump & repair the slot if any of the following occur
+/// * Our version is dead
+/// * Our version is popular pruned and unplayed
+/// * If `epoch_slots_frozen_hash` does not match our local frozen hash (additionally notify fork
+///   choice of duplicate `slot`)
+fn check_epoch_slots_hash_against_bank_status(
     state_changes: &mut Vec<ResultingStateChange>,
     slot: Slot,
     epoch_slots_frozen_hash: Hash,
-    bank_frozen_hash: Hash,
-    is_dead: bool,
+    bank_status: BankStatus,
     is_popular_pruned: bool,
 ) {
-    if epoch_slots_frozen_hash != bank_frozen_hash {
-        if is_dead {
-            // If the cluster duplicate confirmed some version of this slot, then
-            // there's another version of our dead slot
+    match bank_status {
+        BankStatus::Frozen(bank_frozen_hash) if bank_frozen_hash == epoch_slots_frozen_hash => {
+            // Matches, nothing to do
+            return;
+        }
+        BankStatus::Frozen(bank_frozen_hash) => {
+            // The epoch slots hash does not match our frozen hash.
+            warn!(
+                "EpochSlots sample returned slot {} with hash {}, but our version
+                has hash {:?}",
+                slot, epoch_slots_frozen_hash, bank_frozen_hash
+            );
+            if !is_popular_pruned {
+                // If the slot is not already pruned notify fork choice to mark as invalid
+                state_changes.push(ResultingStateChange::MarkSlotDuplicate(bank_frozen_hash));
+            }
+        }
+        BankStatus::Dead => {
+            // Cluster sample found a hash for our dead slot, we must have the wrong version
             warn!(
                 "EpochSlots sample returned slot {} with hash {}, but we marked slot dead",
                 slot, epoch_slots_frozen_hash
             );
-        } else if is_popular_pruned {
+        }
+        BankStatus::Unprocessed => {
+            if !is_popular_pruned {
+                // If the slot is not pruned, it will eventually be replayed so nothing to do now
+                return;
+            }
```

Comment:
Hmm was this actually doing the wrong thing before this change then? Because before this change we would have pushed `ResultingStateChange::RepairDuplicateConfirmedVersion` for `Unprocessed` right?

I guess the idea is if the bank is Unprocessed, we shouldn't get a epoch_slots_frozen_hash from ancestor hashes? Unless there's a bug and ancestor hashes pushes the same signal twice, which I wouldn't be surprised, and we should handle that case

Reply:
I actually just put this in here for completeness, currently there's no way this early return can be triggered. In `on_epoch_slots_frozen` we only let `BankStatus::Unprocessed` proceed to the `check_epoch_slots_hash_against_bank_status` if `is_popular_pruned`.

If it's more readable I could remove this case here, and add a comment to the function doc saying that all `BankStatus::Unprocessed` passed to this function must have `is_popular_pruned` as a precondition.

Reply:
actually it would be caught even further up the chain, `self.can_be_further_replayed()` is always `true` for `BankStatus::Unprocessed` unless `is_popular_pruned` is `true` as well, causing this to early return at `into_state_changes`

Reply:
> If it's more readable I could remove this case here, and add a comment to the function doc saying that all BankStatus::Unprocessed passed to this function must have is_popular_pruned as a precondition.

Yeah, maybe would be good to assert here this can't happen